### PR TITLE
Implemented client version metric

### PR DESCRIPTION
**Pull request by patrickbrophy**
This PR addresses issue #1483. This PR adds a new middleware to collect and parse the user-agent from client requests and increment a counter metric with the label being the client version.

### CODE REVIEW

File: director/director.go [hunk 1]
```diff
@@ -1156,15 +1156,33 @@ func getHealthTestFile(ctx *gin.Context) {
 	}
 }
 
+func collectClientVersionMiddleware(c *gin.Context) {
```
Comment (jhiemstrawisc) [inline director/director.go:1159]: Why add this to a middleware instead of collecting the metric directly in `redirectToCache` and `redirectToOrigin`? My concern is that by the `ShortcutMiddleware` route we have set up in `launchers/director_serve.go` is what gets hit in the majority of client requests (but not cache requests), and it looks like that'll bypass this middleware.

File: director/director.go [hunk 2]
```diff
@@ -1156,15 +1156,33 @@ func getHealthTestFile(ctx *gin.Context) {
...
+	userAgentSlc := c.Request.Header["User-Agent"]
+	if len(userAgentSlc) < 1 {
+		log.Error("Failed to parse User-Agent")
+	}
+
+	userAgent := userAgentSlc[0]
+
+	uaRegExp := regexp.MustCompile(`^pelican-[^\/]+\/\d+\.\d+\.\d+`)
+	if matches := uaRegExp.MatchString(userAgent); !matches {
+		log.Error("Failed to match User-Agent for Pelican client version")
```
Comment (jhiemstrawisc) [inline director/director.go:1169]: Since we use this same regex in other parts of the Director, let's pull it out and functionalize it to guarantee that we can't update one but forget the other.

A second note - I don't know if this really constitutes an "error" because we often use non-pelican clients to interact with the federation as well, such as `curl`. After reading the Cloudflare article you sent, I think we want to avoid creating any prom labels based on an arbitrary incoming UA header, but we should think about how to make this flexible enough to handle the case where the incoming UA isn't Pelican. Maybe it's enough to have a single label that tracks non-pelican clients?

Lastly, w.r.t. cardinality explosion and Pelican headers, do you have a sense of the "cost" of tracking the 56 versions we've already released, knowing that we release between 3-5 new versions every month?

File: metrics/director.go [hunk 3]
```diff
@@ -81,4 +81,9 @@ var (
 		Name: "pelican_director_stat_total",
 		Help: "The total stat queries the director issues. The status can be Succeeded, Cancelled, Timeout, Forbidden, or UnknownErr",
 	}, []string{"server_name", "server_url", "server_type", "result"}) // status: see enums for DirectorStatResult
+
+	PelicanDirectorClientVersionTotal = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pelican_director_client_version_total",
+		Help: "The total amount of requests from client versions.",
```
Comment (jhiemstrawisc) [inline metrics/director.go:87]: Let's say "number" instead of "amount" -- just like "fewer" vs "less", amount describes continuous quantities.

File: utils/utils.go [hunk 4]
```diff
@@ -143,3 +144,15 @@ func ExtractAndMaskIP(ipStr string) (maskedIP string, ok bool) {
 		return ApplyIPMask(ipStr)
 	}
 }
+
+func ExtractVersionAndServiceFromUserAgent(userAgent string) (reqVer, service string) {
```
Comment (jhiemstrawisc) [inline utils/utils.go:148]: Can you add a top-level comment to the function that describes its intended use along with the contents of the outputs (i.e. what reqVer and service look like when they're returned)? Also, can you add a unit test with any edge cases you can think of?

File: director/director.go [hunk 5]
```diff
@@ -1156,6 +1156,33 @@ func getHealthTestFile(ctx *gin.Context) {
 	}
 }
 
+func collectClientVersionMetric(c *gin.Context) {
```
Comment (jhiemstrawisc) [inline director/director.go:1159]: Can you add a top-level comment describing how the function works and what it's supposed to do in the case of non-pelican clients?

File: utils/utils.go [hunk 6]
```diff
@@ -143,3 +144,15 @@ func ExtractAndMaskIP(ipStr string) (maskedIP string, ok bool) {
...
+	uaRegExp := regexp.MustCompile(`^pelican-[^\/]+\/\d+\.\d+\.\d+`)
+	if matches := uaRegExp.MatchString(userAgent); !matches {
+		return "", ""
```
Comment (jhiemstrawisc) [inline utils/utils.go:151]: Maybe we can create a standard "non-pelican" client version for recording?

**Review by jhiemstrawisc — CHANGES_REQUESTED** [review]

**Review by jhiemstrawisc — CHANGES_REQUESTED** [review]
This is getting very close. My last cleanup requests are:

1. Malicious users can still pass mostly-arbitrary user agent headers to Prometheus, potentially leading to cardinality issues. For example, for my local director I can do `curl -v -k -A "pelican-cache/77.123123123.345345345" https://`hostname`:8445/some/object`, and the metric is recorded. Let's make the assumption that the Director will only record versions "other", `[7.0.0, director's version]` and maybe something like "pelican-future" to indicate that the request UA header has a version newer than the director (without the patch version, as you've already done). You can get the Director's version from `config.GetVersion()`, but you'll have to create `version.Version` object to perform comparisons. See `director.versionCompatCheck` for an example of version comparison. I actually think the `pelican-future` version metric is interesting, because it can quickly tell a federation administrator when the majority of client requests are coming from newer versions of pelican -- a good sign that it's time to update your central services! One thing to note is that the only reason I can't get through with a user agent like `pelican-foobar/7.0.0` is that the `versionCompatCheck` errors because foobar is an unrecognized service. If we ever collect this metric in a Director route that doesn't go through `redirectToCache/Origin`, however, this may also be recorded.

2. Right now, both `redirectToCache` and `redirectToOrigin` do a `versionCompatCheck` and then a `collectClientVersionMetric`, but a lot of the guts in these two functions are shared (and can be deduplicated). My proposal: instead of passing the gin context to both the version check and the metric collection, call `ExtractVersionAndServiceFromUserAgent` directly on the gin context and get the version/service. Then pass these values to the version check and metric collection functions. If you agree with point 3, you'll pass both the version and the service to the metric function, and you can validate the input there.

3. This is one is something to think about -- when we record the version metric, should we also record the service as a metric label? For example, what if I want to know "of the Director's traffic, how much is coming from user clients, and how much is coming from caches, and how much is coming from origins?" If we think this is something to record (I lean in favor of yes, even if it bumps the cardinality a bit), I'm in favor of recording this in a separate label in the version metric, so that we can get answers to those types of questions through direct promql queries, instead of having to post process the results.

**Comment by haoming29** [comment]
I'll leave this to @jhiemstrawisc for the final round of review as he has more context to the PR than me at this point, but I hope the changes should be minor

**Review by jhiemstrawisc — APPROVED** [review]
LGTM. Thanks for all the work on this, @patrickbrophy!
